src/helpers/ring.rs

```diff
@@ -26,183 +27,270 @@ pub enum HelperAddr {
-                        // with the same type and destination, we simply panic and abort this task
-                        let buf = &mut *buf.lock().unwrap();
-                        match buf.entry((item.source, item.type_id)) {
-                            Entry::Occupied(_entry) => {
-                                panic!("Duplicated message {item:?}")
-                            }
-                            Entry::Vacant(entry) => entry.insert(item.payload),
-                        };
-                    }
-                }
-            });
+        fn ring_channel(&self, protocol_id: ProtocolId) -> Self::ChannelType {
+            let table = &mut *self.routing_table.lock().unwrap();
 
-            Self {
-                input_queue: tx,
-                left: None,
-                right: None,
-                buf,
-            }
-        }
-
-        fn set_left(&mut self, left: Sender<MessageEnvelope>) {
-            self.left = Some(left);
-        }
+            // Take ownership of the receivers - if they've been created already, move them out
+            // otherwise create them.
+            let rx = match table.active_routes.entry(protocol_id) {
+                Entry::Occupied(mut entry) => entry.get_mut().inbound.take().unwrap(),
+                Entry::Vacant(entry) => entry.insert(Route::new()).inbound.take().unwrap(),
+            };
 
-        fn set_right(&mut self, right: Sender<MessageEnvelope>) {
-            self.right = Some(right);
+            TestChannel {
+                protocol_id,
+                table: Arc::clone(&self.routing_table),
+                senders: HashMap::from([
+                    (HelperAddr::Left, self.left.clone()),
+                    (HelperAddr::Right, self.right.clone()),
+                ]),
+                receivers: rx,
+            }
         }
     }
 
     #[async_trait]
-    impl Ring for TestHelper {
-        async fn send<T: Message>(&self, dest: HelperAddr, msg: T) -> Result<(), Error> {
-            assert!(self.left.is_some());
-            assert!(self.right.is_some());
-
+    impl Ring for TestChannel {
+        async fn send<T: Message>(&mut self, dest: HelperAddr, msg: T) -> Result<(), Error> {
             // inside the envelope we store the sender of the message (i.e. source)
             // but this method accepts the destination. To obtain source from destination
             // we invert it - message send to the left helper is originated from helper on the
             // right side.
-            let (target, source) = match dest {
-                HelperAddr::Left => (self.left.as_ref().unwrap(), HelperAddr::Right),
-                HelperAddr::Right => (self.right.as_ref().unwrap(), HelperAddr::Left),
+            let source = match dest {
+                HelperAddr::Left => HelperAddr::Right,
+                HelperAddr::Right => HelperAddr::Left,
             };
```

Comment:
This is better as a method on HelperAddr.

Reply:
ack, I would still hide it behind `cfg(test)` because it is only used in one place but more importantly I am not sure whether it makes sense in the long term to have it - likely we would have communication patterns that require sending messages to a specific helper, determining source by destination won't work in this case